codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java

```diff
@@ -348,6 +354,75 @@ protected final void discardSomeReadBytes() {
         }
     }
 
+    /**
+     * This method attempts to reclaim space that has already been consumed by the application but still allocated in
+     * the cumulation buffer. By default this method will allocate a new cumulation buffer which will avoid corrupting
+     * any existing views on the cumulation buffer (e.g. {@link ByteBuf#slice()}) and not modifying the underlying
+     * indexes.
+     * @param allocator The {@link ByteBufAllocator} to use if allocation of a new {@link ByteBuf} is necessary.
+     */
+    protected final void tryDiscardSomeReadBytes(ByteBufAllocator allocator) {
+        // Avoid using discardSomeReadBytes because this will modify the underlying storage of the ByteBuf.
+        // Modifying the underlying storage of the ByteBuf means that any slices or views on this data will become
+        // corrupted. This is particularly a problem when offloading user code to a non-EventLoop thread to avoid
+        // blocking the EventLoop. In this case the users code will be executed asynchronously, which maybe a slice
+        // of this cumulation, and if we use discardSomeReadBytes the user code will see corrupted data. If we attempt
+        // to check the ReferenceCount == 1 to see if discardSomeReadBytes is safe to call this may result in the
+        // cumulation buffer growing very large in the case the asynchronous access pattern is persistent.
+        if (cumulation != null && cumulation.readerIndex() >= cumulation.capacity() >>> 1) {
```

Comment:
@Scottmitch should we just set `cumulation` to `null` if `cumulation.isReadable() == false` ? I think there is no need to do allocate anything in this case.

Reply:
This condition is checked in `channelRead()` and the behavior would be different than what was done in `discardSomeReadBytes` so I'm not convinced it is necessary here. This may benefit calls from `channelReadComplete()` but presumably the channel read finally block should catch the condition.

Reply:
I still think it can’t harm

Reply:
Agreed it won't harm but it is redundant (unless I'm missing something). `channelReadComplete()` generally doesn't consume data so I'm assuming why it didn't previously need to account for `!cumulation.isReadable()` (because this condition would be caught in the prior `channelRead()` call, see below). I updated the java doc to clarify the method may `null` out the `cumulation` to be clear it is allowed, but can you clarify which use case you had in mind where we would not clear/reset the `cumulation`? 

```java
// channelRead() code:
                if (cumulation != null && !cumulation.isReadable()) {
                    numReads = 0;
                    releaseCumulation(); //-> release the cumulation here if it has been fully drained
                } else if (++ numReads >= discardAfterReads) {
                    // We did enough reads already try to discard some bytes so we not risk to see a OOME.
                    // See https://github.com/netty/netty/issues/4275
                    numReads = 0;
                    tryDiscardSomeReadBytes(ctx.alloc());
                }
```